File: Sources/ImageHub/Views/BuildHistoryView.swift
import SwiftUI

struct BuildHistoryView: View {
    @EnvironmentObject private var appState: AppState
    @State private var expandedID: UUID?

    var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: 14) {
                if appState.history.isEmpty {
                    EmptyStateView(
                        symbol: "clock.arrow.circlepath",
                        title: "No builds yet",
                        message: "Every drive you write shows up here with its full log, so you can see exactly what happened on a machine you handed over last week."
                    ) {
                        Button("Build USB Drive") {
                            appState.startBuild(template: nil)
                        }
                        .buttonStyle(.borderedProminent)
                        .disabled(appState.templates.templates.isEmpty)
                    }
                    .frame(minHeight: 320)
                } else {
                    ForEach(appState.history) { job in
                        HistoryCard(
                            job: job,
                            isExpanded: expandedID == job.id,
                            onToggle: { expandedID = expandedID == job.id ? nil : job.id }
                        )
                    }
                }
            }
            .padding(.horizontal, 22)
            .padding(.bottom, 26)
        }
        .safeAreaInset(edge: .top, spacing: 0) {
            PageHeader(
                title: "Build History",
                subtitle: appState.history.isEmpty
                    ? "Nothing built in this session"
                    : "\(appState.history.count) build\(appState.history.count == 1 ? "" : "s") this session"
            ) {
                if !appState.history.isEmpty {
                    Button("Clear") {
                        appState.history.removeAll { !$0.isRunning }
                    }
                    .disabled(appState.history.allSatisfy { $0.isRunning })
                }
            }
            .background(.bar)
        }
    }
}

struct HistoryCard: View {
    @ObservedObject var job: BuildJob
    let isExpanded: Bool
    let onToggle: () -> Void

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            Button(action: onToggle) {
                HStack {
                    BuildSummaryRow(job: job)
                    Image(systemName: isExpanded ? "chevron.up" : "chevron.down")
                        .font(.caption)
                        .foregroundStyle(.tertiary)
                        .padding(.trailing, 14)
                }
                .contentShape(Rectangle())
            }
            .buttonStyle(.plain)

            if isExpanded {
                VStack(alignment: .leading, spacing: 8) {
                    Divider()

                    if let started = job.startedAt {
                        Text("Started \(started.briefFormatted)")
                            .font(.caption)
                            .foregroundStyle(.secondary)
                    }

                    ScrollView {
                        VStack(alignment: .leading, spacing: 1) {
                            ForEach(job.log) { line in
                                Text(line.text)
                                    .font(.system(size: 10.5).monospaced())
                                    .foregroundStyle(line.isError ? Color.red : Color.secondary)
                                    .frame(maxWidth: .infinity, alignment: .leading)
                            }
                        }
                        .padding(8)
                    }
                    .frame(height: 220)
                    .background(.quaternary.opacity(0.3), in: RoundedRectangle(cornerRadius: 7))

                    HStack {
                        Button("Copy Log") {
                            NSPasteboard.general.clearContents()
                            NSPasteboard.general.setString(job.logText, forType: .string)
                            ToastCenter.shared.show("Build log copied")
                        }
                        .controlSize(.small)

                        Button("Save Log…") { saveLog() }
                            .controlSize(.small)

                        Spacer()
                    }
                }
                .padding(.horizontal, 14)
                .padding(.bottom, 13)
            }
        }
        .glassCard(padding: 0)
    }

    private func saveLog() {
        let stamp = (job.startedAt ?? Date()).formatted(
            .iso8601.year().month().day().dateSeparator(.dash)
        )
        guard let url = Panels.save(
            title: "Save Build Log",
            suggestedName: "ImageHub-\(job.templateName)-\(stamp).log".replacingOccurrences(of: "/", with: "-"),
            types: [.plainText]
        ) else { return }
        do {
            try Data(job.logText.utf8).write(to: url)
            ToastCenter.shared.show("Log saved", detail: url.lastPathComponent)
        } catch {
            ToastCenter.shared.show(
                "Couldn't save the log",
                detail: error.localizedDescription,
                style: .error
            )
        }
    }
}

File: Windows/App/Ui/BuildHistoryView.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Windows;
using System.Windows.Controls;
using ImageHub.Models;
using ImageHub.Services;
using ImageHub.Support;
using ImageHub.ViewModels;

namespace ImageHub.Views;

/// <summary>
/// Build History: every drive written in this session, each with its full log.
///
/// The log is the point. A machine handed over last week that turns out wrong is
/// diagnosed from what the build actually did, not from what the template says now.
/// </summary>
public sealed class BuildHistoryView : UserControl, IRefreshable
{
    private readonly AppState _state = AppState.Shared;
    private readonly StackPanel _body = new();
    private readonly MainWindow _window;
    private Guid? _expanded;

    public BuildHistoryView(MainWindow window)
    {
        _window = window;
        Content = Ui.Page(
            Ui.PageHeader("Build History", null,
                Ui.Button("Clear", () =>
                {
                    _state.ClearFinishedHistory();
                    Refresh();
                })),
            Ui.Scroll(_body));
        Refresh();
    }

    public void Refresh()
    {
        _body.Children.Clear();

        if (_state.History.Count == 0)
        {
            _body.Children.Add(Ui.EmptyState(
                Glyphs.Circle,
                "No builds yet",
                "Every drive you write shows up here with its full log, so you can see exactly what "
                + "happened on a machine you handed over last week.",
                Ui.Button("Build USB Drive", () => _window.StartBuild(null), "AccentButton")));
            return;
        }

        foreach (BuildJob job in _state.History)
        {
            _body.Children.Add(HistoryCard(job));
        }
    }

    private UIElement HistoryCard(BuildJob job)
    {
        bool expanded = _expanded == job.Id;

        Border summary = BuildRow.Card(job, () =>
        {
            _expanded = expanded ? null : job.Id;
            Refresh();
        });
        summary.Margin = new Thickness(0, 0, 0, expanded ? 0 : 6);
        if (expanded) { summary.CornerRadius = new CornerRadius(8, 8, 0, 0); }

        if (!expanded) { return summary; }

        (Border host, TextBox box) = Ui.LogPane(240);
        box.Text = job.LogText;
        box.TextChanged += (_, _) => box.ScrollToEnd();
        box.ScrollToEnd();

        var detail = new List<UIElement?>
        {
            job.StartedAt is null ? null : Ui.Caption("Started " + Formatting.Brief(job.StartedAt.Value)),
            job.FailureMessage.Length == 0
                ? null
                : Ui.Banner(BannerKind.Error, "Build failed", job.FailureMessage),
            host,
            Ui.Row(8,
                Ui.Button("Copy Log", () =>
                {
                    try
                    {
                        Clipboard.SetText(job.LogText);
                        Notifier.Banner("Build log copied");
                    }
                    catch (Exception error)
                    {
                        Notifier.Banner("Couldn't copy", error.Message, BannerKind.Error);
                    }
                }),
                Ui.Button("Save Log…", () => SaveLog(job))),
        };

        Border body = Ui.Card(Ui.Column(9, detail.ToArray()), 14);
        body.CornerRadius = new CornerRadius(0, 0, 8, 8);
        body.BorderThickness = new Thickness(1, 0, 1, 1);
        body.Margin = new Thickness(0, 0, 0, 6);

        return Ui.Column(0, summary, body);
    }

    private void SaveLog(BuildJob job)
    {
        string stamp = (job.StartedAt ?? DateTime.Now).ToString("yyyy-MM-dd-HHmm");
        string name = $"ImageHub-{job.TemplateName}-{stamp}.log"
            .Replace('/', '-')
            .Replace('\\', '-');
        var dialog = new Microsoft.Win32.SaveFileDialog
        {
            Title = "Save Build Log",
            FileName = name,
            Filter = "Log file (*.log)|*.log|Text file (*.txt)|*.txt",
            InitialDirectory = AppPaths.Logs,
        };
        if (dialog.ShowDialog(_window) != true) { return; }
        try
        {
            File.WriteAllText(dialog.FileName, job.LogText);
            Notifier.Banner("Log saved", Path.GetFileName(dialog.FileName));
        }
        catch (Exception error)
        {
            Notifier.Banner("Couldn't save the log", error.Message, BannerKind.Error);
        }
    }
}
